lib/SPIRV/OCLUtil.cpp

```diff
@@ -777,7 +777,7 @@ bool isSamplerTy(Type *Ty) {
   return STy && STy->hasName() && STy->getName() == kSPR2TypeName::Sampler;
 }
 
-bool isPipeBI(const StringRef MangledName) {
+bool isPipeOrAddressSpaceCastBI(const StringRef MangledName) {
```

Comment:
This is simply a list of functions that are exempt from name mangling, so I don't think that splitting this up into multiple functions would aid clarity. If we were to split this up, that would be an artificial distinction from the point of view of the name mangling. Perhaps you could suggest a better name that avoids mentioning pipes or address space casts altogether, if you think that would help?

Reply:
From name mangling point of view the exempt builtins are grouped at `isNonMangledOCLBuiltin` function. And the list includes also names from `isEnqueueKernelBI` and `isKernelQueryBI`.
I'd consider subgroups builtins in this function as a bug.
I also don't think that splitting the list is artificial, because there are other usages of `isEnqueueKernelBI`, `isKernelQueryBI`  and `isPipeBI`.
Why we should mangle pipe builtins in other way than enqueue kernels builtins for example ? I mean, can we use `isNonMangledOCLBuiltin` instead of `isPipeBI` ?

Reply:
> From name mangling point of view the exempt builtins are grouped at `isNonMangledOCLBuiltin` function. And the list includes also names from `isEnqueueKernelBI` and `isKernelQueryBI`.
> I'd consider subgroups builtins in this function as a bug.
> I also don't think that splitting the list is artificial, because there are other usages of `isEnqueueKernelBI`, `isKernelQueryBI` and `isPipeBI`.
> Why we should mangle pipe builtins in other way than enqueue kernels builtins for example ? I mean, can we use `isNonMangledOCLBuiltin` instead of `isPipeBI` ?

The only other uses of `isPipeBI()` are `mangleBuiltin()`, which calls `manglePipeOrAddressSpaceCastBuiltin()`, and `manglePipeBuiltin()` itself, in its opening assertion.

Maybe we should use `isNonMangledOCLBuiltin()` instead of `isPipeBI()` in `mangleBuiltin()`—is that what you are suggesting? I'm of the impression that this code path wouldn't be exercised for the enqueue kernel and kernel query builtins, but perhaps it would be cleaner.

The `{work,sub}_group_{reserve,commit}_{read,write}_pipe` functions all take a pipe argument, and if I understand correctly, they shouldn't be name mangled. Do you also mean to include the workgroup builtins in your statement? As far as I can see, the semantics of the functions listed in `isPipeBI()` are not of any particular interest here—the conventions for name mangling are the real concern. Of course, the subgroup and workgroup function names begin with `kOCLBuiltinName::WorkGroupPrefix` and `kOCLBuiltinName::SubGroupPrefix`, which leaves me wondering how if they are wrongly handled in e.g. `OCL20ToSPIRV::visitCallInst()` (i.e. handled as ordinary group operations).

I could certainly split the function up, but then what of `manglePipeBuiltin()`—if `isPipeOrAddressSpaceCastBI()` isn't okay, surely `manglePipeOrAddressSpaceCastBuiltin()` needs splitting up too, except that `manglePipeBuiltin()` and `mangleAddressSpaceCastBuiltin()` would then be identical, and we'd have unnecessary code for distinguishing between them.

I would be happy to make changes in this area, although it's not clear to me yet if `OCLUtil::isPipeBI() || OCLUtil::isAddressSpaceCastBI()` would be any cleaner than `OCLUtil::isPipeOrAddressSpaceCastBI()`. My view here is that we might just as well split up the pipe functions that take read pipes from those that take write pipes—the distinction is not a relevant one for the purposes of the code in question.

Reply:
I took a closer look at builtins which should not be mangled at SPIRV->LLVM translation and I found it *messy*. We handle pipe, enqueue kernel, query kernel, and cast builtins in 3 different ways.
So I think this patch is OK for the current implementation.

Reply:
Thanks, @AlexeySotkin. I agree, it is _messy_! However, I'm not arguing that my patch is perfect, just that it's not clear _to me_ what's best, so again, I'd be happy to make changes if you would like, but just wanted to confirm first.